File: app/src/main/java/com/smoothradio/radio/feature/radio_list/ui/adapter/RadioListRecyclerViewAdapter.java
package com.smoothradio.radio.feature.radio_list.ui.adapter;

import static com.smoothradio.radio.service.StreamService.StreamStates.BUFFERING;
import static com.smoothradio.radio.service.StreamService.StreamStates.PLAYING;
import static com.smoothradio.radio.service.StreamService.StreamStates.PREPARING;

import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.ads.AdRequest;
import com.smoothradio.radio.MainActivity;
import com.smoothradio.radio.R;
import com.smoothradio.radio.core.model.AdItem;
import com.smoothradio.radio.core.model.ListItem;
import com.smoothradio.radio.core.model.RadioStation;
import com.smoothradio.radio.databinding.AdviewBinding;
import com.smoothradio.radio.databinding.EmptyFavouritiesBinding;
import com.smoothradio.radio.databinding.RadioitemBinding;
import com.smoothradio.radio.feature.radio_list.util.StationDiffUtilCallback;
import com.smoothradio.radio.feature.radio_list.util.StationSortHelper;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;


public class RadioListRecyclerViewAdapter extends RecyclerView.Adapter {
    // Constants for view types
    private static final int AD_VIEW = 0;
    private static final int ITEM_VIEW = 1;
    private static final int EMPTY_ITEM = 2;

    // Constants for Advertisement Positions
    private static final int AD_POSITION_1 = 1;
    private static final int AD_POSITION_2 = 9;
    private static final int AD_POSITION_3 = 36;
    private static final int AD_POSITION_4 = 45;
    private static final int AD_POSITION_5 = 54;
    private static final int AD_POSITION_6 = 72;

    // Data collections
    public final List<RadioStation> favouriteList = new ArrayList<>();
    private final List<ListItem> filteredStationNameList = new ArrayList<>();

    private final List<RadioStation> stationList;
    private final List<ListItem> recyclerViewItems;

    // State-related fields
    private final RadioStation emptyRadioStation = new RadioStation(0, 0, "", "", "", "", false, false);
    private DisplayState currentState = DisplayState.POPULAR;
    private String state = "";

    // Dependencies
    private final MainActivity.RadioStationActionHandler radioStationActionHandler;

    public RadioListRecyclerViewAdapter(List<RadioStation> radioStations, MainActivity.RadioStationActionHandler radioStationActionHandler) {
        this.radioStationActionHandler = radioStationActionHandler;

        stationList = new ArrayList<>(radioStations);

        List<ListItem> displayList = stationList.isEmpty()
                ? new ArrayList<>()
                : injectAdItems(stationList);


        recyclerViewItems = new ArrayList<>(displayList);

    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        switch (viewType) {
            case ITEM_VIEW:
                RadioitemBinding radioitemBinding = RadioitemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
                return new ItemViewViewHolder(radioitemBinding);
            case AD_VIEW:
                AdviewBinding adviewBinding = AdviewBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
                return new AdViewHolder(adviewBinding);
            default:
                EmptyFavouritiesBinding binding = EmptyFavouritiesBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
                return new EmptyFavouriteListViewHolder(binding);
        }
    }


    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ListItem item = recyclerViewItems.get(position);

        if (holder instanceof ItemViewViewHolder && item instanceof RadioStation) {
            bindRadioStation((ItemViewViewHolder) holder, (RadioStation) item);
        } else if (holder instanceof AdViewHolder && item instanceof AdItem) {
            bindAdView((AdViewHolder) holder);
        }
    }

    private void bindRadioStation(@NonNull ItemViewViewHolder holder, @NonNull RadioStation radioStation) {

        // Favorite icon
        int favIconRes = radioStation.isFavorite()
                ? R.drawable.favorite_20px_filled
                : R.drawable.favorite_20px;
        holder.binding.ivFavourite.setImageResource(favIconRes);
        // Basic UI
        holder.binding.ivLogo.setBackgroundResource(radioStation.getLogoResource());
        holder.binding.tvChannelName.setText(radioStation.getStationName());
        holder.binding.tvFrequency.setText(radioStation.getFrequency());
        holder.binding.tvLocation.setText(radioStation.getLocation());

        // Play button click
        holder.binding.ivPlay.setImageResource(R.drawable.playicon);
        holder.binding.ivPlay.setOnClickListener(new PlayOnclickListener(radioStation));
        holder.binding.ivFavourite.setOnClickListener(new FavouriteListener(radioStation, holder));

        // Animation on scroll
        holder.binding.clLayout.setAnimation(
                AnimationUtils.loadAnimation(holder.itemView.getContext(), R.anim.recyclerviewscrollanimation)
        );

        // Handle playing states
        updatePlayerUI(holder, radioStation);
    }

    private void updatePlayerUI(@NonNull ItemViewViewHolder holder, @NonNull RadioStation station) {
        boolean isPlaying = station.isPlaying();

        if (!isPlaying) {
            holder.binding.lottieLoadingAnimation.setVisibility(View.INVISIBLE);
            holder.binding.ivPlay.setVisibility(View.VISIBLE);
            holder.binding.ivPlay.setImageResource(R.drawable.playicon);
            return;
        }

        if (Objects.equals(state, PLAYING)) {
            holder.binding.lottieLoadingAnimation.setVisibility(View.INVISIBLE);
            holder.binding.ivPlay.setVisibility(View.VISIBLE);
            holder.binding.ivPlay.setImageResource(R.drawable.pauseicon);
        } else if (Objects.equals(state, PREPARING) || Objects.equals(state, BUFFERING)) {
            holder.binding.lottieLoadingAnimation.setVisibility(View.VISIBLE);
            holder.binding.ivPlay.setVisibility(View.INVISIBLE);
        } else {
            holder.binding.lottieLoadingAnimation.setVisibility(View.INVISIBLE);
            holder.binding.ivPlay.setVisibility(View.VISIBLE);
            holder.binding.ivPlay.setImageResource(R.drawable.playicon);
        }
    }

    private void bindAdView(@NonNull AdViewHolder holder) {
        AdRequest adRequest = new AdRequest.Builder().build();
        holder.binding.adViewRecyclerviewItem.loadAd(adRequest);
    }


    @Override
    public int getItemViewType(int position) {
        ListItem item = recyclerViewItems.get(position);

        // Empty states
        if ((currentState.equals(DisplayState.SEARCH) && filteredStationNameList.isEmpty()) ||
                (currentState.equals(DisplayState.FAVORITES) && favouriteList.isEmpty())) {
            return EMPTY_ITEM;
        }
        if (item instanceof AdItem) {
            return AD_VIEW;
        }
        return ITEM_VIEW;
    }


    @Override
    public int getItemCount() {
        return recyclerViewItems.size();
    }


    class ItemViewViewHolder extends RecyclerView.ViewHolder {
        private final RadioitemBinding binding;

        public ItemViewViewHolder(@NonNull RadioitemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }


    public class AdViewHolder extends RecyclerView.ViewHolder {
        private final AdviewBinding binding;

        public AdViewHolder(@NonNull AdviewBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    class EmptyFavouriteListViewHolder extends RecyclerView.ViewHolder {
        EmptyFavouritiesBinding binding;

        public EmptyFavouriteListViewHolder(@NonNull EmptyFavouritiesBinding binding) {
            super(binding.getRoot());
        }
    }

    class PlayOnclickListener implements View.OnClickListener {
        private int stationId;
        private RadioStation radioStation;

        private PlayOnclickListener(RadioStation radioStation) {
            this.radioStation = radioStation;
            this.stationId = radioStation.getId();
        }

        @Override
        public void onClick(View view) {
            radioStationActionHandler.onStationSelected(radioStation);
            setPlayingStation(stationId);
        }
    }


    class FavouriteListener implements View.OnClickListener {
        RadioStation radioStation;
        ItemViewViewHolder itemViewViewHolder;

        FavouriteListener(RadioStation radioStation, ItemViewViewHolder itemViewViewHolder) {
            this.radioStation = radioStation;
            this.itemViewViewHolder = itemViewViewHolder;
        }

        @Override
        public void onClick(View view) {
            boolean isFavorite = radioStation.isFavorite();
            radioStation.setFavorite(!isFavorite);

            radioStationActionHandler.onToggleFavorite(radioStation, !isFavorite);

            // Update UI icon
            itemViewViewHolder.binding.ivFavourite.setImageResource(
                    isFavorite ? R.drawable.favorite_20px : R.drawable.favorite_20px_filled
            );

//            handlePostFavoriteAction();
        }

        private void handlePostFavoriteAction() {
            radioStationActionHandler.onRequestHideKeyboard();

            if (currentState.equals(DisplayState.FAVORITES)) {
                sortFavourites();
            }
        }

    }

    public RadioStation getStationAtPosition(int position) {
        ListItem item = recyclerViewItems.get(position);
        if (item instanceof RadioStation) {
            return (RadioStation) item;
        }
        return emptyRadioStation;
    }


    public int getPositionOfStation(int stationId) {
        RadioStation lastRadioStation = new RadioStation(stationId, 0, "", "", "", "", true, false);
        return recyclerViewItems.indexOf(lastRadioStation);
    }

    public void updateStation(RadioStation station) {
        int position = recyclerViewItems.indexOf(station);
        if (position != RecyclerView.NO_POSITION) {
            notifyItemChanged(position);
        }
    }

    public boolean listIsEmpty() {
        return recyclerViewItems.isEmpty();
    }

    public void setState(String state) {
        this.state = state;
    }

    public void setPlayingStation(int stationId) {

        List<RadioStation> updatedList = new ArrayList<>();

        for (RadioStation station : stationList) {
            RadioStation copy = new RadioStation(station);
            copy.setPlaying(station.getId() == stationId);
            updatedList.add(copy);
        }

        update(updatedList);
    }

    public void setFavouriteStations(List<RadioStation> favouriteStations) {
        Log.d("RadioListRecyclerViewAdapter", "setFavouriteStations: " + favouriteStations.size() );
        favouriteList.clear();
        favouriteList.addAll(favouriteStations);
        if (currentState == DisplayState.FAVORITES) {
//            sortFavourites();
        }
    }

    public void update(List<RadioStation> newList) {
        stationList.clear();
        stationList.addAll(newList);

        if (currentState == DisplayState.FAVORITES) return;

        List<ListItem> oldList = new ArrayList<>(recyclerViewItems);
        List<ListItem> newListWithAds = injectAdItems(newList);

        if (recyclerViewItems.isEmpty()) {
            recyclerViewItems.addAll(newListWithAds);
            notifyDataSetChanged();
            return;
        }

        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(
                new StationDiffUtilCallback(oldList, newListWithAds)
        );

        recyclerViewItems.clear();
        recyclerViewItems.addAll(newListWithAds);

        diffResult.dispatchUpdatesTo(this);
    }

    public void updateFavorites(List<RadioStation> newFavorites) {
        favouriteList.clear();
        favouriteList.addAll(newFavorites);

        if (currentState == DisplayState.POPULAR) return;

        List<ListItem> oldList = new ArrayList<>(recyclerViewItems);
        List<ListItem> newFavoriteList = new ArrayList<>(newFavorites);

        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(
                new StationDiffUtilCallback(oldList, newFavoriteList)
        );

        recyclerViewItems.clear();
        recyclerViewItems.addAll(newFavoriteList);

        diffResult.dispatchUpdatesTo(this);
    }

    private List<ListItem> injectAdItems(List<RadioStation> originalList) {
        List<ListItem> modifiedList = new ArrayList<>(originalList);

        int[] adPositions = {AD_POSITION_1, AD_POSITION_2, AD_POSITION_3, AD_POSITION_4, AD_POSITION_5, AD_POSITION_6};
        for (int pos : adPositions) {
            if (pos <= modifiedList.size()) {
                modifiedList.add(pos, new AdItem());
            }
        }
        return modifiedList;
    }


    //--------------------
    //Search
    //-------------------
    public void filter(String input) {
        // Reset to popular if currently in special modes
        if (currentState != DisplayState.POPULAR && currentState != DisplayState.SEARCH) {
            sortPopular();
        }
        // Clear previous results
        filteredStationNameList.clear();

        if (TextUtils.isEmpty(input)) {
            handleEmptySearch();
        } else {
            handleSearchQuery(input);
        }

        updateDisplayedResults();
    }

    private void handleEmptySearch() {
        currentState = DisplayState.POPULAR;

        List<ListItem> withAds = injectAdItems(stationList);


        filteredStationNameList.addAll(withAds);
    }

    private void handleSearchQuery(String query) {
        currentState = DisplayState.SEARCH;
        String lowerQuery = query.toLowerCase();

        for (RadioStation station : stationList) {
            if (station.getStationName().toLowerCase().contains(lowerQuery)) {
                filteredStationNameList.add(station);
            }
        }
    }

    private void updateDisplayedResults() {
        recyclerViewItems.clear();

        if (filteredStationNameList.isEmpty()) {
            recyclerViewItems.add(emptyRadioStation);
        } else {
            recyclerViewItems.addAll(filteredStationNameList);
        }

        notifyDataSetChanged();
    }

    //--------------------
    //Sorting Logic
    //-------------------
    public void sortAndDisplay(DisplayState state) {
        currentState = state;

        // Copy and sort
        List<RadioStation> sortedList = new ArrayList<>(stationList);

        if (state == DisplayState.ASCENDING) {
            StationSortHelper.sortByName(sortedList, true);
        } else if (state == DisplayState.DESCENDING) {
            StationSortHelper.sortByName(sortedList, false);
        }

        // Inject ads
        List<ListItem> withAds = injectAdItems(sortedList);


        recyclerViewItems.clear();
        recyclerViewItems.addAll(withAds);

        notifyDataSetChanged();
    }

    public void sortPopular() {
        currentState = DisplayState.POPULAR;

        List<ListItem> withAds = injectAdItems(stationList);

        recyclerViewItems.clear();
        recyclerViewItems.addAll(withAds);

        notifyDataSetChanged();
    }

    public void sortFavourites() {
        currentState = DisplayState.FAVORITES;

        if (favouriteList.isEmpty()) {
            showNoFavoritesMessage();
            displayEmptyState();
        } else {
            displaySortedFavorites();
        }
    }

    private void showNoFavoritesMessage() {
        radioStationActionHandler.onRequestshowToast("No favorite stations added");
    }

    private void displayEmptyState() {
        recyclerViewItems.clear();
        recyclerViewItems.add(emptyRadioStation);
        notifyDataSetChanged();
    }

    private void displaySortedFavorites() {
        recyclerViewItems.clear();
        recyclerViewItems.addAll(favouriteList);

        notifyDataSetChanged();
    }

    public interface RadioStationActionListener {
        void onStationSelected(RadioStation station);

        void onToggleFavorite(RadioStation station, boolean isFavorite);

        void onRequestHideKeyboard();

        void onRequestshowToast(String message);
    }

    public enum DisplayState {POPULAR, FAVORITES, ASCENDING, DESCENDING, SEARCH}
}

File: app/src/main/java/com/smoothradio/radio/feature/radio_list/ui/adapter/RadioListRecyclerViewAdapter.kt
package com.smoothradio.radio.feature.radio_list.ui.adapter

import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.google.android.gms.ads.AdRequest
import com.smoothradio.radio.MainActivity
import com.smoothradio.radio.R
import com.smoothradio.radio.core.model.AdItem
import com.smoothradio.radio.core.model.ListItem
import com.smoothradio.radio.core.model.RadioStation
import com.smoothradio.radio.databinding.AdviewBinding
import com.smoothradio.radio.databinding.EmptyFavouritiesBinding
import com.smoothradio.radio.databinding.RadioitemBinding
import com.smoothradio.radio.feature.radio_list.util.StationDiffUtilCallback
import com.smoothradio.radio.feature.radio_list.util.StationSortHelper
import com.smoothradio.radio.service.StreamService.StreamStates.BUFFERING
import com.smoothradio.radio.service.StreamService.StreamStates.PLAYING
import com.smoothradio.radio.service.StreamService.StreamStates.PREPARING

class RadioListRecyclerViewAdapter(
    radioStations: List<RadioStation>,
    private val radioStationActionHandler: MainActivity.RadioStationActionHandler
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val AD_VIEW = 0
        private const val ITEM_VIEW = 1
        private const val EMPTY_ITEM = 2

        private const val AD_POSITION_1 = 1
        private const val AD_POSITION_2 = 9
        private const val AD_POSITION_3 = 36
        private const val AD_POSITION_4 = 45
        private const val AD_POSITION_5 = 54
        private const val AD_POSITION_6 = 72
    }

    private val favouriteList = mutableListOf<RadioStation>()
    private val filteredStationNameList = mutableListOf<ListItem>()

    private val stationList = radioStations.toMutableList()
    private val recyclerViewItems: MutableList<ListItem> = mutableListOf()

    private val emptyRadioStation = RadioStation(0, 0, "", "", "", "", false, false)
    private var currentState = DisplayState.POPULAR
    private var state: String = ""

    init {
        val displayList = if (stationList.isEmpty()) emptyList() else injectAdItems(stationList)
        recyclerViewItems.addAll(displayList)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            ITEM_VIEW -> {
                val binding =
                    RadioitemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
                ItemViewViewHolder(binding)
            }

            AD_VIEW -> {
                val binding =
                    AdviewBinding.inflate(LayoutInflater.from(parent.context), parent, false)
                AdViewHolder(binding)
            }

            else -> {
                val binding = EmptyFavouritiesBinding.inflate(
                    LayoutInflater.from(parent.context),
                    parent,
                    false
                )
                EmptyFavouriteListViewHolder(binding)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = recyclerViewItems[position]
        when {
            holder is ItemViewViewHolder && item is RadioStation -> bindRadioStation(holder, item)
            holder is AdViewHolder && item is AdItem -> bindAdView(holder)
        }
    }

    private fun bindRadioStation(holder: ItemViewViewHolder, radioStation: RadioStation) =
        with(holder.binding) {
            ivFavourite.setImageResource(
                if (radioStation.isFavorite) R.drawable.favorite_20px_filled else R.drawable.favorite_20px
            )

            ivLogo.setBackgroundResource(radioStation.logoResource)
            tvChannelName.text = radioStation.stationName
            tvFrequency.text = radioStation.frequency
            tvLocation.text = radioStation.location

            ivPlay.setImageResource(R.drawable.playicon)
            ivPlay.setOnClickListener(PlayOnclickListener(radioStation))
            ivFavourite.setOnClickListener(FavouriteListener(radioStation, holder))

            if (holder.bindingAdapterPosition == RecyclerView.NO_POSITION) {
                clLayout.animation = AnimationUtils.loadAnimation(
                    holder.itemView.context,
                    R.anim.recyclerviewscrollanimation
                )
            }

            updatePlayerUI(holder, radioStation)
        }


    private fun updatePlayerUI(holder: ItemViewViewHolder, station: RadioStation) =
        with(holder.binding) {
            when {
                !station.isPlaying -> {
                    lottieLoadingAnimation.isInvisible = true
                    ivPlay.isVisible = true
                    ivPlay.setImageResource(R.drawable.playicon)
                }

                state == PLAYING -> {
                    lottieLoadingAnimation.isInvisible = true
                    ivPlay.isVisible = true
                    ivPlay.setImageResource(R.drawable.pauseicon)
                }

                state == PREPARING || state == BUFFERING -> {
                    lottieLoadingAnimation.isVisible = true
                    ivPlay.isInvisible = true
                }

                else -> {
                    lottieLoadingAnimation.isInvisible = true
                    ivPlay.isVisible = true
                    ivPlay.setImageResource(R.drawable.playicon)
                }
            }
        }


    private fun bindAdView(holder: AdViewHolder) {
        val adRequest = AdRequest.Builder().build()
        holder.binding.adViewRecyclerviewItem.loadAd(adRequest)
    }

    override fun getItemViewType(position: Int): Int {
        val item = recyclerViewItems[position]
        return when {
            (currentState == DisplayState.SEARCH && filteredStationNameList.isEmpty()) ||
                    (currentState == DisplayState.FAVORITES && favouriteList.isEmpty()) -> EMPTY_ITEM

            item is AdItem -> AD_VIEW
            else -> ITEM_VIEW
        }
    }

    override fun getItemCount(): Int = recyclerViewItems.size

    inner class ItemViewViewHolder(val binding: RadioitemBinding) :
        RecyclerView.ViewHolder(binding.root)

    inner class AdViewHolder(val binding: AdviewBinding) : RecyclerView.ViewHolder(binding.root)
    inner class EmptyFavouriteListViewHolder(val binding: EmptyFavouritiesBinding) :
        RecyclerView.ViewHolder(binding.root)

    inner class PlayOnclickListener(private val radioStation: RadioStation) : View.OnClickListener {
        private val stationId = radioStation.id
        override fun onClick(view: View?) {
            radioStationActionHandler.onStationSelected(radioStation)
            setPlayingStation(stationId)
        }
    }

    inner class FavouriteListener(
        private val radioStation: RadioStation,
        private val holder: ItemViewViewHolder
    ) : View.OnClickListener {
        override fun onClick(view: View?) {
            val newFavoriteState = !radioStation.isFavorite
            radioStation.isFavorite = newFavoriteState
            radioStationActionHandler.onToggleFavorite(radioStation, newFavoriteState)

            holder.binding.ivFavourite.setImageResource(
                if (newFavoriteState) R.drawable.favorite_20px_filled else R.drawable.favorite_20px
            )
        }

    }

    fun getStationAtPosition(position: Int): RadioStation {
        val item = recyclerViewItems[position]
        return if (item is RadioStation) item else emptyRadioStation
    }

    fun getPositionOfStation(stationId: Int): Int {
        val dummyStation = RadioStation(stationId, 0, "", "", "", "", true, false)
        return recyclerViewItems.indexOf(dummyStation)
    }

    fun updateStation(station: RadioStation) {
        val position = recyclerViewItems.indexOf(station)
        if (position != RecyclerView.NO_POSITION) notifyItemChanged(position)
    }

    fun listIsEmpty(): Boolean = recyclerViewItems.isEmpty()

    fun setState(state: String) {
        this.state = state
    }

    fun setPlayingStation(stationId: Int) {
        val updatedList = stationList.map { station ->
            station.copy(isPlaying = station.id == stationId)
        }
        update(updatedList)
    }

    private var lastSearchQuery: String? = null


    fun update(newList: List<RadioStation>) {
        stationList.clear()
        stationList.addAll(newList)

        if (currentState == DisplayState.FAVORITES) return

        val displayList: List<ListItem> = when (currentState) {
            DisplayState.SEARCH -> {
                val query = lastSearchQuery.orEmpty().lowercase()
                val refreshedList = stationList.filter {
                    it.stationName.lowercase().contains(query)
                }
                filteredStationNameList.clear()
                filteredStationNameList.addAll(refreshedList)
                filteredStationNameList
            }


            DisplayState.ASCENDING ->
                injectAdItems(stationList.sortedBy { it.stationName.lowercase() })

            DisplayState.DESCENDING ->
                injectAdItems(stationList.sortedByDescending { it.stationName.lowercase() })

            else -> injectAdItems(stationList)
        }

        if (recyclerViewItems.isEmpty()) {
            recyclerViewItems.addAll(displayList)
            notifyDataSetChanged()
            return
        }

        val diff = DiffUtil.calculateDiff(StationDiffUtilCallback(recyclerViewItems, displayList))
        recyclerViewItems.clear()
        recyclerViewItems.addAll(displayList)
        diff.dispatchUpdatesTo(this)
    }



    fun updateFavorites(newFavorites: List<RadioStation>) {
        favouriteList.clear()
        favouriteList.addAll(newFavorites)

        if (currentState != DisplayState.FAVORITES) return

        val newList = if (favouriteList.isEmpty()) listOf(emptyRadioStation) else favouriteList
        val diff = DiffUtil.calculateDiff(StationDiffUtilCallback(recyclerViewItems, newList))
        recyclerViewItems.clear()
        recyclerViewItems.addAll(newList)
        diff.dispatchUpdatesTo(this)
    }

    private fun injectAdItems(originalList: List<RadioStation>): List<ListItem> {
        val modifiedList = originalList.toMutableList<ListItem>()
        val adPositions = listOf(
            AD_POSITION_1,
            AD_POSITION_2,
            AD_POSITION_3,
            AD_POSITION_4,
            AD_POSITION_5,
            AD_POSITION_6
        )
        for (pos in adPositions) {
            if (pos <= modifiedList.size) modifiedList.add(pos, AdItem())
        }
        return modifiedList
    }

    fun filter(input: String?) {
        if (currentState !in listOf(DisplayState.POPULAR, DisplayState.SEARCH)) {
            sortPopular()
        }

        lastSearchQuery = input.orEmpty().trim()
        filteredStationNameList.clear()

        if (lastSearchQuery!!.isEmpty()) {
            handleEmptySearch()
        } else {
            handleSearchQuery(lastSearchQuery!!)
        }

        updateDisplayedResults()
    }


    private fun handleEmptySearch() {
        currentState = DisplayState.POPULAR
        val withAds = injectAdItems(stationList)
        filteredStationNameList.addAll(withAds)
    }

    private fun handleSearchQuery(query: String) {
        currentState = DisplayState.SEARCH
        val lowerQuery = query.lowercase()
        stationList.forEach { station ->
            if (station.stationName.lowercase().contains(lowerQuery)) {
                filteredStationNameList.add(station)
            }
        }
    }

    private fun updateDisplayedResults() {
        recyclerViewItems.clear()
        if (filteredStationNameList.isEmpty()) {
            recyclerViewItems.add(emptyRadioStation)
        } else {
            recyclerViewItems.addAll(filteredStationNameList)
        }
        notifyDataSetChanged()
    }

    fun sortAndDisplay(state: DisplayState) {
        currentState = state
        val sortedList = ArrayList(stationList)
        when (state) {
            DisplayState.ASCENDING -> StationSortHelper.sortByName(sortedList, true)
            DisplayState.DESCENDING -> StationSortHelper.sortByName(sortedList, false)
            else -> {}
        }
        val withAds = injectAdItems(sortedList)
        recyclerViewItems.clear()
        recyclerViewItems.addAll(withAds)
        notifyDataSetChanged()
    }

    fun sortPopular() {
        currentState = DisplayState.POPULAR
        val withAds = injectAdItems(stationList)
        recyclerViewItems.clear()
        recyclerViewItems.addAll(withAds)
        notifyDataSetChanged()
    }

    fun sortFavourites() {
        currentState = DisplayState.FAVORITES
        recyclerViewItems.clear()

        if (favouriteList.isEmpty()) {
            radioStationActionHandler.onRequestshowToast("No favorite stations added")
            recyclerViewItems.add(emptyRadioStation)
        } else {
            recyclerViewItems.addAll(favouriteList)
        }

        notifyDataSetChanged()
    }

    interface RadioStationActionListener {
        fun onStationSelected(station: RadioStation)
        fun onToggleFavorite(station: RadioStation, isFavorite: Boolean)
        fun onRequestHideKeyboard()
        fun onRequestshowToast(message: String)
    }

    enum class DisplayState { POPULAR, FAVORITES, ASCENDING, DESCENDING, SEARCH }
}
